cores/esp8266/WString.cpp

```diff
@@ -682,14 +682,13 @@ int String::lastIndexOf(char ch) const {
 int String::lastIndexOf(char ch, unsigned int fromIndex) const {
     if (fromIndex >= len())
         return -1;
-    char *writeTo = wbuffer();
-    char tempchar = writeTo[fromIndex + 1]; // save the replaced character
-    writeTo[fromIndex + 1] = '\0';
-    char *temp = strrchr(writeTo, ch);
-    writeTo[fromIndex + 1] = tempchar; // restore character
-    if (temp == NULL)
-        return -1;
-    return temp - writeTo;
+    for (unsigned int index = fromIndex; index > 0; --index) {
+        if (buffer()[index] == ch)
+            return index;
+    }
+    if (buffer()[0] == ch)
```

Comment:
note that the loop could include this condition as well (...also seen in std::string::rfind...)
```cpp
#include <array>
#include <iostream>

const char x[] {
    'a', 'b', 'c', 'd', 'e'
};

int main() {
    auto size = std::size(x);
    std::cout << "size is " << size << '\n';
    if (size) {
        --size;
        for (++size; size--;) {
            std::cout << x[size] << ':' << size << '\n';
        }
    }
}
```
```
size is 5
e:4
d:3
c:2
b:1
a:0
```

Reply:
:+1: then the not-found case can also be integrated in the loop
```cpp
    int index = fromIndex + 1;
    while (--index >= 0 && buffer()[index] != ch);
    return index;
```